pkg/cmd/recycle/walker.go

```diff
@@ -0,0 +1,131 @@
+package recycle
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+)
+
+type Walker struct {
+	walkFn filepath.WalkFunc
+
+	// fsuid holds the last fsuid we set
+	fsuid int64
+
+	// lstat is for testing, defaults to os.Lstat
+	lstat func(name string) (fi os.FileInfo, err error)
+
+	// setfsuid is for testing, defaults to syscall.Setfsuid
+	setfsuid func(uid int) (err error)
+
+	// getuid is for testing, defaults to fileinfo.Sys().(*syscall.Stat_t).Uid
+	getuid func(os.FileInfo) (int64, error)
+}
+
+func NewWalker(walkFn filepath.WalkFunc) *Walker {
+	return &Walker{
+		walkFn:   walkFn,
+		fsuid:    -1,
+		lstat:    os.Lstat,
+		setfsuid: setfsuid,
+		getuid:   getuid,
+	}
+}
+
+func (w *Walker) Walk(root string) error {
+	// Lock threads, so our Setfsuid calls always apply to the same thread
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// The launching process must have stat permission on the root dir
+	info, err := w.lstat(root)
+	if err != nil {
+		return w.walkFn(root, info, err)
+	}
+	return w.walk(root, info)
+}
+
+func (w *Walker) walk(path string, info os.FileInfo) error {
+	var err error
+
+	// Descend first
+	if info.IsDir() {
+		// switch to the dir's owner
+		err = w.becomeOwner(info)
+		if err != nil {
+			return w.walkFn(path, info, err)
+		}
+
+		// read dir info
+		names, err := readDirNames(path)
+		if err != nil {
+			return w.walkFn(path, info, err)
+		}
+
+		// visit files
+		for _, name := range names {
+			filename := filepath.Join(path, name)
+			fileInfo, err := w.lstat(filename)
+			if err != nil {
+				if err := w.walkFn(filename, fileInfo, err); err != nil {
```

Comment:
all errors are passed through `walkFn` to propagate them or cancel them. `walk` will only return an errors `walkFn` returned